functions/src/etl/SurveyManager.ts

```diff
@@ -0,0 +1,122 @@
+		{ name: '05-checkin-4', startDateOffsetMonths: 24 },
+		{ name: '06-checkin-5', startDateOffsetMonths: 30 },
+		{ name: '07-offboarding', startDateOffsetMonths: 36 },
+		{ name: '08-offboarded-checkin-1', startDateOffsetMonths: 42 },
+		{ name: '09-offboarded-checkin-2', startDateOffsetMonths: 48 },
+		{ name: '10-offboarded-checkin-3', startDateOffsetMonths: 60 },
+		{ name: '11-offboarded-checkin-4', startDateOffsetMonths: 72 },
+	];
+
+	constructor(firestoreAdmin: FirestoreAdmin, authAdmin: AuthAdmin) {
+		this.firestoreAdmin = firestoreAdmin;
+		this.authAdmin = authAdmin;
+	}
+
+	/**
+	 * Batch implementation to create all surveys for all recipients.
+	 * Checks for existing surveys and ignores them. So, it can be called multiple times without creating duplicates.
+	 */
+	createAllSurveys = functions.https.onCall(async (_, { auth }) => {
+		await this.firestoreAdmin.assertGlobalAdmin(auth?.token?.email);
+
+		const recipients = await this.firestoreAdmin.collection<Recipient>(RECIPIENT_FIRESTORE_PATH).get();
+
+		await Promise.all(
+			recipients.docs
+				.filter((recipient) => recipient.data().progr_status != RecipientProgramStatus.Waitlisted)
+				.map(async (recipient) => this.createAllSurveysForRecipient(recipient))
+		);
+	});
+
+	/**
+	 * Create all surveys for 1 recipient. TODO trigger this method when a recipient gets selected.
+	 * Checks for existing surveys and ignores them. So, it can be called multiple times without creating duplicates.
+	 */
+	createAllSurveysForRecipient = async (recipient: FirebaseFirestore.QueryDocumentSnapshot<Recipient>) => {
+		if (recipient.data().si_start_date) {
+			return Promise.all(
+				this.surveysToCreate.map(async (survey) => {
+					const dueDate = moment(recipient.data().si_start_date).add(survey.startDateOffsetMonths, 'M');
+					const surveyStatus = dueDate.isBefore(moment()) ? SurveyStatus.Missed : SurveyStatus.New;
+					await this.createSurvey(
+						recipient.id,
+						survey.name,
+						recipient.data().first_name,
+						recipient.data().main_language,
+						surveyStatus,
+						dueDate.toDate()
+					);
+				})
+			);
+		} else {
+			console.log('No start date for recipient ${}');
+			return Promise.resolve();
+		}
+	};
+	/**
+	 * First checks if the survey already exists and return without any action if so.
+	 * Otherwise, create a new auth user specifically for this survey.
+	 * This will be used in the firestore security rules to check if someone can update a survey.
+	 */
+	createSurvey = async (
+		recipientId: string,
+		surveyName: string,
+		recipientName: string,
+		language: RecipientMainLanguage,
+		status: SurveyStatus,
+		due_date_at: Date
+	) => {
+		const surveysCollection = this.firestoreAdmin.collection<Survey>(
+			[RECIPIENT_FIRESTORE_PATH, recipientId, SURVEY_FIRETORE_PATH].join('/')
+		);
+		const existingSurveys = await surveysCollection.listDocuments();
+		const surveyExists = existingSurveys.filter((s) => s.id === surveyName).length > 0;
```

Comment:
Since you know the full path of the document, you can do this a little easier, imo.

See:
https://github.com/socialincome-san/public/blob/fcb42bc8af539f19f2a8171351d3fdd4a82831db/functions/src/admin/AdminPaymentTaskProcessor.ts#L63-L75

Reply:
Thanks for the pointers! Better, will change